jaxtyping/_decorator.py

```diff
@@ -309,6 +310,29 @@ def __init__(self, *args, **kwargs):
             # in which case make a best-effort attempt to add shape information for any
             # type errors.
 
+            # we want to detect generators, and ignore return annotations on them,
+            # to avoid issues with O(n) typechecking trying to typecheck yielded values
+            wrp = fn
+            while hasattr(wrp, "__wrapped__"):
+                wrp = wrp.__wrapped__
+
+            if inspect.isgeneratorfunction(wrp) or inspect.isasyncgenfunction(wrp):
+                # recursively parse all the annotations, and mark all the jaxtyping
+                # annotations as not needing instance checks, while still being
+                # visible as original ones for the typechecker
+                def modify_annotation(ann):
+                    if isinstance(ann, _MetaAbstractArray):
+                        setattr(ann, "_skip_instancecheck", True)
+                    elif hasattr(ann, "__args__"):
+                        for sub_ann in get_args(ann):
+                            modify_annotation(sub_ann)
+                    elif hasattr(ann, "__origin__"):
+                        modify_annotation(get_origin(ann))
```

Comment:
In contrast this one should probably technically be
```python
origin = get_origin(ann)
if origin is not None:
    modify_annotation(origin)
```
(IIRC `__args__` and `__origin__` are actually implementation details that we're not really supposed to touch.)